### DISCUSSION TITLE
Despite using an efficient approach to upload files, the POST endpoint still seems to be "blocking"...

### DISCUSSION THREAD

**Discussion by NB-123** [fastapi/fastapi#11363] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request, HTTPException, status
from streaming_form_data import StreamingFormDataParser
from streaming_form_data.targets import FileTarget, ValueTarget
from streaming_form_data.validators import MaxSizeValidator
import streaming_form_data
from datetime import datetime
from starlette.requests import ClientDisconnect
import os

MAX_FILE_SIZE = 1024 * 1024 * 1024 * 4  # = 4GB
MAX_REQUEST_BODY_SIZE = MAX_FILE_SIZE + 1024

app = FastAPI()

class MaxBodySizeException(Exception):
    def __init__(self, body_len: str):
        self.body_len = body_len

class MaxBodySizeValidator:
    def __init__(self, max_size: int):
        self.body_len = 0
        self.max_size = max_size

    def __call__(self, chunk: bytes):
        self.body_len += len(chunk)
        if self.body_len > self.max_size:
            raise MaxBodySizeException(body_len=self.body_len)
 
@app.post('/upload')
async def upload(request: Request):
    body_validator = MaxBodySizeValidator(MAX_REQUEST_BODY_SIZE)
    filename = request.headers.get('Filename')
    
    if not filename:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, 
            detail='Filename header is missing')
    try:
        filepath = os.path.join('./', os.path.basename(filename)) 
        file_ = FileTarget(filepath, validator=MaxSizeValidator(MAX_FILE_SIZE))
        data = ValueTarget()
        parser = StreamingFormDataParser(headers=request.headers)
        parser.register('file', file_)
        parser.register('data', data)
        
        async for chunk in request.stream():
            body_validator(chunk)
            parser.data_received(chunk)
    except ClientDisconnect:
        print("Client Disconnected")
    except MaxBodySizeException as e:
        raise HTTPException(status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, 
           detail=f'Maximum request body size limit ({MAX_REQUEST_BODY_SIZE} bytes) exceeded ({e.body_len} bytes read)')
    except streaming_form_data.validators.ValidationError:
        raise HTTPException(status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE, 
            detail=f'Maximum file size limit ({MAX_FILE_SIZE} bytes) exceeded') 
    except Exception:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, 
            detail='There was an error uploading the file') 
   
    if not file_.multipart_filename:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail='File is missing')

    print(data.value.decode())
    print(file_.multipart_filename)
    print(f"[{datetime.now()}] [upload] Successfuly uploaded {file_.multipart_filename}")
    return {"message": f"Successfuly uploaded {filename}"}

@app.get("/file/{file_id}")
async def get_data(file_id: str):
    return retrieve_file_data(file_id)

def retrieve_file_data(file_id):
    print(f"[{datetime.now()}] [retrieve_file_data] Starting file retrieval")
    # MORE CODE FOLLOWS
```

### Description

I have an endpoint called /upload in which I can post files to. I followed this [great answer](https://stackoverflow.com/questions/73442335/how-to-upload-a-large-file-≥3gb-to-fastapi-backend/73443824?noredirect=1#comment137912642_73443824) on stackoverflow to write this endpoint. After uploading a file, the user calls the GET method to get some data about the file. 

I have dockerized this app and here are the logs: 
2024-03-26 11:13:09 [2024-03-26 10:13:09.284038] [upload] Successfuly uploaded test.pdf
2024-03-26 11:13:09 INFO:     192.168.65.1:20194 - "POST /extractor/upload HTTP/1.1" 200 OK
2024-03-26 11:16:01 [2024-03-26 10:13:09.298668] [retrieve_file_data] Starting file retrieval

As you can see, the upload and the retrieve_file_data functions run very quickly after each other. However, looking a the locker logs, there is a delay between the POST and when the next function runs (3 minutes of delay!). I am not sure what is going on or how to solve this so would appreciate some guidance. I suspect it has to do with the inner workings of FastAPI. I do not seem to have any blocking functions or code either.

### Operating System

macOS

### FastAPI Version

0.110.0

### Pydantic Version

1.10.14

### Python Version

3.10.7

**Comment by YuriiMotov** [comment]
Does your `def retrieve_file_data(file_id)`contain heavy code?
You use `async def` for endpoint, but inside you have not async code. You should either use `def` instead of `async def` for `get_data` or run `retrieve_file_data` in the threadpool.

FastAPI runs all the endpoints declared with `async def` in one thread (one event loop). And if one of them have heavy not async code, others will have to wait.

**Comment by YuriiMotov** [comment]
Closing this. Feel free to reopen it when you are ready to provide more info
